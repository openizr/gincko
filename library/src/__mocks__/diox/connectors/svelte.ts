/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { readable } from 'svelte/store/index';

/**
 * diox/connectors/svelte mock.
 */
const connect = vi.fn(() => {
  const state = {
    steps: [{ id: 'start' }, { id: 'end' }],
    loading: process.env.LOADING === 'true',
  };
  return vi.fn(() => readable(state, (set) => {
    set(state);
  }));
});

export default connect;
