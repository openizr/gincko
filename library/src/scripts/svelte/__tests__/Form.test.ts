/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @jest-environment jsdom
 */

import Form from 'scripts/svelte/__tests__/TestForm.svelte';
import { render, fireEvent, createEvent } from '@testing-library/svelte';

jest.mock('scripts/core/Engine');
jest.mock('diox/connectors/svelte');
jest.mock('scripts/svelte/Step.svelte');

describe('svelte/Form', () => {
  const configuration = {
    id: 'test',
    root: 'start',
    steps: {
      start: {
        fields: {},
      },
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
    delete process.env.LOADING;
  });

  test('renders correctly - loading next step', async () => {
    process.env.LOADING = 'true';
    const { container } = render(Form, { props: { configuration } });
    expect(container.firstChild).toMatchSnapshot();
    const { container: newContainer } = render(Form, { props: { configuration } });
    expect(newContainer.firstChild).toMatchSnapshot();
  });

  test('renders correctly - with active step', async () => {
    const { container } = render(Form, { props: { configuration, activeStep: 'start' } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('prevents native form submission', async () => {
    const { container } = render(Form, { props: { configuration } });
    const form = container.getElementsByTagName('form')[0];
    const event = createEvent.submit(form);
    event.preventDefault = jest.fn();
    await fireEvent(form, event);
    expect(event.preventDefault).toHaveBeenCalled();
  });
});
