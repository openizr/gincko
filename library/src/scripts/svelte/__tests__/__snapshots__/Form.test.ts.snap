// Vitest Snapshot v1

exports[`svelte/Form > renders correctly - loading next step 1`] = `
<div>
  <form
    class="gincko"
    id="test"
  >
    <div
      class="gincko__steps"
    >
      
      <div
        id="step"
      >
        {"step":{"id":"start"},"index":0,"isActive":false,"customComponents":{}}
      </div>
      <!--&lt;Step&gt;-->
      
      <div
        id="step"
      >
        {"step":{"id":"end"},"index":1,"isActive":true,"customComponents":{}}
      </div>
      <!--&lt;Step&gt;-->
       
      <div
        slot="loader"
      >
        LOADING...
      </div>
    </div>
  </form>
  <!--&lt;Form&gt;-->
  <!--&lt;TestForm&gt;-->
</div>
`;

exports[`svelte/Form > renders correctly - loading next step 2`] = `
<div>
  <form
    class="gincko"
    id="test"
  >
    <div
      class="gincko__steps"
    >
      
      <div
        id="step"
      >
        {"step":{"id":"start"},"index":0,"isActive":false,"customComponents":{}}
      </div>
      <!--&lt;Step&gt;-->
      
      <div
        id="step"
      >
        {"step":{"id":"end"},"index":1,"isActive":true,"customComponents":{}}
      </div>
      <!--&lt;Step&gt;-->
       
      <div
        slot="loader"
      >
        LOADING...
      </div>
    </div>
  </form>
  <!--&lt;Form&gt;-->
  <!--&lt;TestForm&gt;-->
</div>
`;

exports[`svelte/Form > renders correctly - with active step 1`] = `
<div>
  <form
    class="gincko"
    id="test"
  >
    <div
      class="gincko__steps"
    >
      
      <div
        id="step"
      >
        {"step":{"id":"start"},"index":0,"isActive":false,"customComponents":{}}
      </div>
      <!--&lt;Step&gt;-->
      
      <div
        id="step"
      >
        {"step":{"id":"end"},"index":1,"isActive":true,"customComponents":{}}
      </div>
      <!--&lt;Step&gt;-->
       
      <div
        slot="loader"
      >
        LOADING...
      </div>
    </div>
  </form>
  <!--&lt;Form&gt;-->
  <!--&lt;TestForm&gt;-->
</div>
`;
